### PR TITLE
Save execution ledger on disk with RocksDB

### PR DESCRIPTION
**Pull request by Eitu33**
https://github.com/massalabs/massa/issues/2579

### CODE REVIEW

**Comment by Eitu33** [comment]
@damip I believe this PR is ready to be merged, here is what I chose to _not_ implement in this one because I believe it should be done in separate PRs:

- Save the bytecode in the datastore: requires a bit of work to get it working, I find it to be out of scope here
- Allow the speculative ledger to lazily query any sub-entry: Did do that to query address-specific entries from the execution history, but implementing it in the speculative ledger requires more work and I also find it to be out of scope
- LedgerDB::delete function: this can be done in this PR, I just need a definition of how we want to implement it

File: massa-ledger/src/ledger.rs
Comment (AurelienFT) [inline massa-ledger/src/ledger.rs:112]: I think we should make a temporary bootstrap state method because otherwise this PR will block launching new testnets and we shouldn't merge something that prevent launching testnets.

File: massa-ledger/src/ledger_db.rs
Comment (AurelienFT) [inline massa-ledger/src/ledger_db.rs:9]: The other DB for consensus is save in the storage folder of `massa-node`. I think we should group them.

File: massa-ledger/src/ledger_db.rs
Comment (AurelienFT) [inline massa-ledger/src/ledger_db.rs:94]: Can we save varint instead ? To have a more light ledger.

File: massa-ledger/src/ledger_db.rs
Comment (AurelienFT) [inline massa-ledger/src/ledger_db.rs:115]: I think we can fetch all the data of an address using prefix.
Here is the docs of rocksdb about it : https://github.com/facebook/rocksdb/wiki/Iterator
And an issue of rust repo that show some examples : https://github.com/rust-rocksdb/rust-rocksdb/issues/577

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:94]: Yeah I'm pretty sure we can.

File: massa-ledger/src/ledger.rs
Comment (damip) [inline massa-ledger/src/ledger.rs:153]: Amount should have its own compact deserialization routine (so that from_raw is never used)

Comment (Eitu33) [reply massa-ledger/src/ledger.rs:153]: This particular one was not supposed to stay, what I'm doing now is: 
```rust
    pub fn get_parallel_balance(&self, addr: &Address) -> Option<Amount> {
        self.sorted_ledger
            .get_entry(addr, LedgerSubEntry::Balance)
            .map(|bytes| {
                Amount::from_raw(
                    u64::from_varint_bytes(&bytes)
                        .expect("critical: invalid balance format")
                        .0,
                )
            })
    }
```

Is that how it should be done?

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:256]: this has linear complexity

note that it can be implemented with constant complexity because active_history elements represent strictly consecutive slots in increasing order

File: massa-api/src/public.rs
Comment (Eitu33) [reply massa-api/src/public.rs:580]: Actually, it has been done.

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:252]: better performance can be achieved with

* finding highest history index for which `output.slot > slot` using constant complexity because we know the 1st slot of the history and we know that the history is composed of strictly consecutive slots
* iterating from that index (non-included) backwards down to 0

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:274]: same remark as above absent vs dont-know + lookup in the final ledger

Comment (damip) [inline massa-execution-worker/src/execution.rs:274]: same remark as above absent vs dont-know + lookup in the final ledger

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:249]: But this function is only used in [get_final_and_active_parallel_balance](https://github.com/massalabs/massa/blob/ledger-unification/massa-execution-worker/src/execution.rs#L835). Searching in the final ledger would mean doing twice the same operation. I think we should either search in the final ledger if `get_active_balance_at_slot` does not succeed and not return the final balance, or the other way around but not both.

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:844]: Note that if the balance is not found in the history, this should do a 2nd lookup into the final_state.
This has a performance impact and should be avoided because we already did this lookup above

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:860]: same remark as above on the performance impact of reading the final ledger potentially twice

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:252]: I don't understand how we can find the highest history index for which `output.slot > slot` given that `active_history` is of type `VecDeque`. I don't think it's possible to achieve constant complexity search with that data structure but maybe I'm wrong?

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:25]: in a future PR, we could fuse Bytecode into the datastore

File: massa-ledger/src/ledger.rs
Comment (Eitu33) [reply massa-ledger/src/ledger.rs:176]: I do not exactly know where the may come from, but from what've I've red (see https://github.com/facebook/rocksdb/issues/4149#issuecomment-405966488) it seems outdated. I used it in the first place to not break the Ledger interface too much, and because it was qualified as more light weight that using `get()`, but I have seen by myself that it is not trustworthy, forgot to add a comment about it.

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:105]: we *always* need to avoid usig to_raw. Here Balance::to_bytes_compact should be used

File: massa-execution-worker/src/execution.rs
Comment (damip) [reply massa-execution-worker/src/execution.rs:249]: Oh ok then I think that this function should be renamed to reflect that we are only looking into the active history (we are not getting the absolute active value).

In turn, when looking for the "active balance", get_final_and_active_parallel_balance should return the info from the active history if available, otherwise it should return the final value

Comment (damip) [reply massa-execution-worker/src/execution.rs:249]: To clarify: "active balance" does not mean "modified within the history". It means "the absolute balance at the output of the current slot". That means that if the active history does not give any info on it, it needs to be returned from the final state.

File: massa-ledger/src/ledger.rs
Comment (damip) [reply massa-ledger/src/ledger.rs:176]: ok let's avoid using it :)

File: massa-ledger/src/ledger.rs
Comment (damip) [reply massa-ledger/src/ledger.rs:153]: from_raw / to_raw should basically never be used (except when we absolutely need it in the SC engine)

This code snippet assumes that u64 is the underlying representation of Amount, which might change.

We should always use Amount::from_bytes_compact and Amount::to_bytes_compact for serialization, and never assume anything about the underlying representation of Amount outside of the Amount module.

Comment (Eitu33) [reply massa-ledger/src/ledger.rs:153]: Oh I forgot there was an implemented trait that already does that, will change it now.

Comment (AurelienFT) [reply massa-ledger/src/ledger.rs:153]: I join here to say that now in main there is the new way to deserialize for amount also with the deserializer. Maybe use it to avoid change in the near future.

File: massa-ledger/src/ledger_db.rs
Comment (damip) [reply massa-ledger/src/ledger_db.rs:105]: the day we change the underlying representation of Amount, we will have to find all the places where we made assumptions instead of using the functions exposed by the Amount module whose role is exactly that.

For the Result, you might just unwrap() it for now with a comment. In the future we will stop using this trait after the work on serialization improvements :)

**Review by damip — CHANGES_REQUESTED** [review]
Nice work !
A few comments need to be addressed before we merge but overall it's good.

Should we discuss rocksDB in-RAM caching in a followup ?

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:44]: Done.

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:39]: as a good practice let's avoid hardcoded "0u8", "1u8" etc... and use constants declared at the top of the file :)

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:46]: the order looks wrong. Let's start with addr, then 1u8 (and not the opposite because it will screw up prefix ranges)

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:194]: Prefix was implemented for the intermediate bootstrap, will use it in the delete function.

File: massa-ledger/src/ledger_db.rs
Comment (damip) [reply massa-ledger/src/ledger_db.rs:46]: it is also incompatible with data_prefix

File: massa-ledger/src/ledger_db.rs
Comment (damip) [reply massa-ledger/src/ledger_db.rs:52]: (for example 2u8)

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:46]: We can change it later but I need it to be in that order for now because of the intermediate bootstrap implementation.

Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:46]: Can add a comment on that.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:52]: Will leave a comment about that but it just extra space taken for nothing right now IMHO.

Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:55]: Done.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs]: Done.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs]: Done.

File: massa-ledger/src/ledger.rs
Comment (Eitu33) [reply massa-ledger/src/ledger.rs:173]: Added a comment about it.

File: massa-ledger/src/ledger.rs
Comment (Eitu33) [reply massa-ledger/src/ledger.rs:18]: This will be discussed in the follow up.

File: massa-ledger/src/ledger.rs
Comment (Eitu33) [reply massa-ledger/src/ledger.rs]: Actually I just removed it.

**Comment by Eitu33** [comment]
@damip All the requests have been resolved, here is what I propose for the follow up PR's:

- Fusing bytecode with datastore (estimated time between 4 and 8 hours)
- Implement lazy sub-entry query for speculative ledger (estimated time between 4 and 8 hours)
- RocksDB cache implementation (didn't investigate enough to estimate)
- Moving blockgraph ledger to massa-ledger (didn't investigate enough to estimate accurately, would say between 4 and 7 days)

What would you like to go for first?

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:27]: Yes I believe that should be part of the follow up as well.

**Comment by Eitu33** [comment]
@damip Apparently RocksDB uses LRU cache by default: https://github.com/facebook/rocksdb/wiki/Block-Cache#lru-cache

**Review by damip — APPROVED** [review]

**Comment by damip** [comment]
> @damip All the requests have been resolved, here is what I propose for the follow up PR's:

Good job ! This is a very important step for the refactoring.

>     * Fusing bytecode with datastore (estimated time between 4 and 8 hours)

Let's keep this on the side for now, we need to discuss it a bit more I think.

>     * Implement lazy sub-entry query for speculative ledger (estimated time between 4 and 8 hours)

That's something you can do now, but make sure to sync with Adrien L and Evgeni to not break the tooling and web3 libs on API changes.

> 
>     * RocksDB cache implementation (didn't investigate enough to estimate)

Apparently that part is solved by rocksdb.

>     * Moving blockgraph ledger to massa-ledger (didn't investigate enough to estimate accurately, would say between 4 and 7 days)

Let's do that last (not sure it will fit in testnet 11)

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:858]: There is still a problem here.

In the case where active_balance = None, it might mean that either:
* we got no info on the entry by looking at the active history => in that case we need to return the final_balance
* OR that the active history deletes the entry => in that case we need to return None instead of the final_balance => the current code does not handle that correctly

File: massa-api/src/public.rs
Comment (damip) [inline massa-api/src/public.rs:551]: We need to document every API change in the API doc, and also ping Evgeni and Adrien because it will break all the tooling (eg. explorer etc...)

**Review by damip — CHANGES_REQUESTED** [review]

File: massa-api/src/public.rs
Comment (Eitu33) [reply massa-api/src/public.rs]: I implemented the precedent behaviour with the new ledger so that it stays compatible.

**Comment by Eitu33** [comment]
@damip ready to be merged!

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:270]: let's make sure the panic in there cannot be triggered by someone asking for the balance at a block in the future through the api

File: massa-execution-worker/src/execution.rs
Comment (damip) [inline massa-execution-worker/src/execution.rs:273]: I think it's

```
let iter = self.active_history.iter().skip(n).rev();
```

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:40]: for now let's make sure we destroy the db before every bootstrap (since we cannot resume yet and we don't want to keep outdated elements from the previous launch in the db)

File: massa-ledger/src/ledger_db.rs
Comment (damip) [inline massa-ledger/src/ledger_db.rs:80]: I am still confused by what this function does exactly :)

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:273]: Are you sure? `active_history` has the newest element A at the back, so if we want to fetch the data of a given slot B the `skip` should happen after the `rev` because we are skipping from A to B, and parcouring the slots that come before B to determine the value. Or am I getting it wrong?

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:270]: This function is currently never used, I left it there as a blueprint for my future implementation of speculative ledger lazy sub-queries. No chance of that happening right now, I noted it and will make sure it cannot happen when it is actually used by the API.

Comment (damip) [reply massa-execution-worker/src/execution.rs:270]: okay

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:273]: FI, this depends on which implementation of `get_active_index` we are using: https://github.com/massalabs/massa/pull/2590#discussion_r883769585

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:258]: Done.

File: massa-execution-worker/src/execution.rs
Comment (Eitu33) [reply massa-execution-worker/src/execution.rs:273]: Done.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:289]: There is no such thing as `write_slot_changes(changes: LedgerChanges, new_slot: Slot)`, only fn `apply(&mut self, changes: LedgerChanges)`. Do we want something like that? If yes for what purpose?

Also I set all the write functions as `pub(crate)`.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:83]: Improved the comment.

File: massa-ledger/src/ledger_db.rs
Comment (Eitu33) [reply massa-ledger/src/ledger_db.rs:40]: Done.

File: massa-ledger/src/ledger_db.rs
Comment (damip) [reply massa-ledger/src/ledger_db.rs:292]: Oh OK I see. 

Basically the idea was to not import WriteBatch nor anything rocksdb-related into ledger.rs

Comment (damip) [reply massa-ledger/src/ledger_db.rs:292]: Also, the slot inside the metadata column should also be updated atomically in the same batch

**Review by damip — APPROVED** [review]

**Comment by damip** [comment]
bors merge

**Comment by Eitu33** [comment]
bors merge

**Review by AurelienFT — APPROVED** [review]
Approving so that bors don't fail.

**Pull request merged**
